src/os/bluestore/ZonedAllocator.cc

```diff
@@ -145,27 +174,92 @@ void ZonedAllocator::init_rm_free(uint64_t offset, uint64_t length) {
   ceph_assert(length % zone_size == 0);
 
   for ( ; length; length -= zone_size) {
-    advance_write_pointer(++zone_num, zone_size);
+    increment_write_pointer(++zone_num, zone_size);
     ldout(cct, 40) << __func__ << " set zone 0x" << std::hex
 		   << zone_num << " write pointer to 0x" << zone_size << dendl;
   }
 }
 
-bool ZonedAllocator::zoned_get_zones_to_clean(std::deque<uint64_t> *zones_to_clean) {
+const std::set<uint64_t> *ZonedAllocator::zoned_get_zones_to_clean(void) const {
+  ldout(cct, 10) << __func__ << dendl;
+  return &zones_to_clean;
+}
+
+bool ZonedAllocator::low_on_space(void) {
+  ldout(cct, 10) << __func__ << dendl;
+
+  ceph_assert(zones_to_clean.empty());
+
+  uint64_t conventional_size = first_seq_zone_num * zone_size;
+  uint64_t sequential_size = size - conventional_size;
+  uint64_t sequential_num_free = num_free - conventional_size;
+  double free_ratio = static_cast<double>(sequential_num_free) / sequential_size;
+
+  ldout(cct, 10) << __func__ << " free size " << sequential_num_free
+		 << " total size " << sequential_size
+		 << " free ratio is " << free_ratio << dendl;
+
   // TODO: make 0.25 tunable
-  if (static_cast<double>(num_free) / size > 0.25) {
+  if (free_ratio > 0.25) {
+    ldout(cct, 10) << __func__ << " no need to clean" << dendl;
     return false;
   }
-  {
-    std::lock_guard l(lock);
-    // TODO: populate |zones_to_clean| with the numbers of zones that should be
-    // cleaned.
-  }
+  ldout(cct, 10) << __func__<< " running out of free space!" << dendl;
   return true;
 }
 
-void ZonedAllocator::zoned_set_zone_states(std::vector<zone_state_t> &&_zone_states) {
+void ZonedAllocator::find_zones_to_clean(void) {
+  ldout(cct, 10) << __func__ << dendl;
+
+  ceph_assert(zones_to_clean.empty());
+
+  // TODO: make this tunable; handle the case when there aren't this many zones
+  // to clean.
+  num_zones_to_clean = 1;
```

Comment:
IMO that should be done at the end when everything is ready for processing in another thread

Reply:
Done.